python/GafferUI/EditMenu.py

```diff
@@ -192,6 +193,72 @@ def paste( menu ) :
 
 		s.graphEditor.frame( s.script.selection(), extend = True )
 
+def duplicateWithInputs( menu ) :
+
+	s = scope( menu )
+
+	errorHandler = GafferUI.ErrorDialogue.ErrorHandler(
+		title = "Errors Occurred During Duplication",
+		closeLabel = "Oy vey",
+		parentWindow = s.scriptWindow
+	)
+
+	with Gaffer.UndoScope( s.script ), errorHandler :
+		source = s.script.serialise( s.parent, s.script.selection() )
+
+		# Paste the selected nodes into an empty script to determine the order
+		# in which they are serialized.
+		dummyNodes = []
+		dummyScript = Gaffer.ScriptNode()
+		caDummy = dummyScript.childAddedSignal().connect( lambda parent, child : dummyNodes.append( child ), scoped = True )
+		dummyScript.execute( source )
+
+		# Paste them again into the destination script, which will create them
+		# in the same order as `dummyScript`, but with different names due to
+		# name collision handling.
+		newNodes = []
+		caScript = s.parent.childAddedSignal().connect( lambda parent, child : newNodes.append( child ), scoped = True )
+		s.script.execute( source, s.parent )
+
+		def duplicateInputs( source, dest ) :
+
+			if isinstance( dest, Gaffer.Plug ) :
+				if dest.direction == Gaffer.Plug.Direction.Out :
+					return
+
+				if source.getInput() is not None and dest.getInput() is None:
+					if isinstance( source.getInput().node(), Gaffer.Expression ) :
+						# If the source input is an expression, we'd be connecting
+						# to a potentially ephemeral plug on the expression that may
+						# change with expression output changes. We connect to the source
+						# directly for better reliability at the expense of consistency.
+						dest.setInput( source )
+					else :
+						dest.setInput( source.getInput() )
+					return
+
+			for i in range( 0, len( source.children() ) ) :
+				duplicateInputs( source[i], dest[i] )
+
+		# The name of a dummy node relative to `dummyScript` will match the
+		# name of the source node relative to the scoped parent from which
+		# to duplicate input connections.
+		for i in range( 0, len( newNodes ) ) :
+			duplicateInputs(
+				s.parent.descendant( dummyNodes[i].relativeName( dummyScript ) ),
```

Comment:
Minor thing, but I think the use of `descendant()` and `relativeName()` here implies that we're potentially iterating nested nodes, but I think we know that they're direct children of `s.parent` and `dummyScript`. So `s.parent[dummyNodes[i].getName()]` should be sufficient and a bit quicker?

Reply:
That would prevent us from connecting a nested node's plug to a plug at or above `s.parent`, correct? The so-to-speak "punching through" plugs I think they've been referred to in some places.

Reply:
I don't think so. Here we're just iterating over the _immediate_ children created by the serialisation. The visiting of nested nodes is done in the recursive `duplicateInputs()` function...at least that's how I understood it...

Reply:
Oh yes, of course. Fixed in 60dc668427f0fe3265e595d415d647521380ddfa